apps/arweave/src/ar_kv.erl

```diff
@@ -11,650 +11,529 @@
-				{ok, Key, _Value} when Key > EndOffsetBinary ->
-					{ok, #{}};
-				{ok, Key, Value} ->
-					get_range4(Iterator, #{ Key => Value }, EndOffsetBinary);
-				{error, invalid_iterator} ->
-					{ok, #{}};
-				{error, Reason} ->
-					{error, Reason}
-			end;
-		Error ->
-			Error
-	end;
-get_range3(DB, StartOffsetBinary, EndOffsetBinary) ->
-	case rocksdb:iterator(DB, [{total_order_seek, true}]) of
-		{ok, Iterator} ->
-			case rocksdb:iterator_move(Iterator, {seek, StartOffsetBinary}) of
-				{ok, Key, _Value} when Key > EndOffsetBinary ->
-					{ok, #{}};
-				{ok, Key, Value} ->
-					get_range4(Iterator, #{ Key => Value }, EndOffsetBinary);
-				{error, invalid_iterator} ->
-					{ok, #{}};
-				{error, Reason} ->
-					{error, Reason}
-			end;
-		Error ->
-			Error
-	end.
 
-delete(Name, Key, RetryCount) ->
-	[{_, Ref}] = ets:lookup(?MODULE, Name),
-	case catch delete2(Ref, Key) of
-		{'EXIT', Exc} ->
-			case RetryCount of
-				0 ->
-					?LOG_ERROR([{event, ar_kv_error}, {op, delete},
-							{key, ar_util:encode(Key)},
-							{exception, io_lib:format("~p", [Exc])}]),
-					{error, Exc};
-				_ ->
-					case reconnect(Name, Ref) of
-						{ok, _Ref2} ->
-							delete(Name, Key, RetryCount - 1);
-						Error ->
-							Error
-					end
+%% @doc Apply callback if it is possible to obtain the iterator for the database.
+%% The callback will get an iterator as an argument.
+with_iterator(Name, Op, IteratorOptions, Callback) ->
+	with_db(Name, Op, fun
+		(#db{db_handle = Db, cf_handle = undefined}) ->
+			case rocksdb:iterator(Db, IteratorOptions) of
+				{ok, Iterator} -> apply(Callback, [Iterator]);
+				{error, IteratorError} -> {error, IteratorError}
 			end;
-		Reply ->
-			Reply
+		(#db{db_handle = Db, cf_handle = Cf}) ->
+			case rocksdb:iterator(Db, Cf, IteratorOptions) of
+				{ok, Iterator} -> apply(Callback, [Iterator]);
+				{error, IteratorError} -> {error, IteratorError}
+			end
+	end).
+
+
+
+%% @doc Apply callback if the database is available.
+%% The callback will get the database record (#db{}) as an argument.
+with_db(Name, Op, Callback) ->
+	try
+		case ets:lookup(?MODULE, Name) of
+			[] ->
+				{error, db_not_found};
+			[DbRec0] ->
+				apply(Callback, [DbRec0])
+		end
+	catch
+		Exc ->
+			?LOG_ERROR([{event, db_operation_failed}, {op, Op},
+				{name, io_lib:format("~p", [Name])},
+				{reason, io_lib:format("~p", [Exc])}]),
+			{error, failed}
 	end.
 
-delete2({DB, CF}, Key) ->
-	rocksdb:delete(DB, CF, Key, []);
-delete2(DB, Key) ->
-	rocksdb:delete(DB, Key, []).
-
-delete_range(Name, Start, End, RetryCount) ->
-	[{_, Ref}] = ets:lookup(?MODULE, Name),
-	case catch delete_range2(Ref, Start, End) of
-		{'EXIT', Exc} ->
-			case RetryCount of
-				0 ->
-					?LOG_ERROR([{event, ar_kv_error}, {op, delete_range},
-							{start_key, ar_util:encode(Start)},
-							{end_key, ar_util:encode(End)},
-							{exception, io_lib:format("~p", [Exc])}]),
-					{error, Exc};
-				_ ->
-					case reconnect(Name, Ref) of
```

Comment:
Yes, but the problem was bigger than that.
The reconnect was called from the functions that were executed on the caller side (not in the `ar_kv` process), so this calls may happen concurrently, a few at once. This means, that if the database for some reason is down, and the db reference is outdated, few other processes will do attempt `gen_server:call` with demand to reconnect the database. This calls will be serialized in the process mailbox, and gen_server will do close/open sequence several times. While the database is closed, other processes will find the database reference dead, will call for reconnects, and this will never end.

Reply:
ah got it - sounds like a mess. Good call removing.

Reply:
> This calls will be serialized in the process mailbox, and gen_server will do close/open sequence several times. While the database is closed, other processes will find the database reference dead, will call for reconnects, and this will never end.

This is not true, the first reconnect changes the reference and the subsequent processes simply take it - https://github.com/ArweaveTeam/arweave/blob/master/apps/arweave/src/ar_kv.erl#L171

Nevertheless, we probably do not need the reconnect functionality now so it makes sense to remove it.

Reply:
@ldmberman yes, that's the race condition, because several processes will hit the same gen_server:call at the same time and the `ar_kv` server will do reconnect sequence several times.
If you're referring to the fact that one database is only used by one process (not sure that this is true), then yes, there is no race condition just because the race contains of one process.